app/src/main/java/com/mamadou/newsapp/MainActivity.kt

```diff
@@ -33,27 +27,79 @@ class MainActivity : AppCompatActivity() {
         setContentView(binding.root)
         binding.articleRecyclerView.run {
             adapter = articleAdapter
+
         }
         fetchArticles()
+
+        val queryTextListener = object : SearchView.OnQueryTextListener {
+            override fun onQueryTextSubmit(query: String?): Boolean {
+                return true
+            }
+
+            override fun onQueryTextChange(newText: String?): Boolean {
+                newText?.let { searchQuery ->
+                    viewsModel.searchArticles(searchQuery)
+                }
+                return true
+            }
+        }
+
+        binding.searchView.setOnQueryTextListener(queryTextListener)
+
+
         binding.swiperefresh.setOnRefreshListener {
             fetchArticles()
         }
     }
 
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menuInflater.inflate(R.menu.menu, menu)
+        val switchItem = menu!!.findItem(R.id.wifi_switch_items)
+        val switch = switchItem.actionView?.findViewById<SwitchCompat>(R.id.wifi_switch)
+        viewModel.isDownloadOverWifiOnly.observe(this) { isDownloadOverWifiOnly ->
+            if (switch != null) {
+                isDownloadOverWifiOnly.also { switch.isChecked = it }
+            }
+            if (isDownloadOverWifiOnly) {
+                if (switch != null) {
+                    switch.text = getString(R.string.switch_on, switchItem.title)
+                }
+            } else {
+                if (switch != null) {
+                    switch.text = getString(R.string.switch_off, switchItem.title)
+                }
+            }
+        }
+        switch?.setOnCheckedChangeListener { _, _ ->
+            viewModel.toggleDownloadOverWifiOnly()
+
+        }
+        return true
+    }
+
+
     private val articleAdapter =
         ArticleRecyclerAdapter { article ->
             val newsDetailIntent = Intent(this@MainActivity, NewsDetailsActivity::class.java)
             newsDetailIntent.putExtra(INTENT_EXTRA_ARTICLE, article)
             startActivity(newsDetailIntent)
         }
 
+    private val viewsModel by viewModels<NewsListViewModel> {
+        NewsListViewModel.Factory(
+            newsRepo = App.newsRepository
+        )
+    }
+
+
 
     private fun fetchArticles() {
-        networkStatusChecker.performIfConnectedToInternet ({
+   //     networkStatusChecker.performIfConnectedToInternet ({
             binding.swiperefresh.isRefreshing = true
-            lifecycleScope.launch(IO) {
-                val result = remoteApi.getArticles()
-                withContext(Main) {
+        //    lifecycleScope.launch(IO) {
+        viewsModel.articles.observe(this) { result ->
```

Comment:
One note here, the observation of `viewModel.articles` should only happen once and can be set up in `onCreate()`.  Instead you should expose a `fetchArticles()` method from the viewmodel so you can call that here instead:

```kotlin
// in your viewmodel refactor the init block like this:
init {
    fetchArticles()
 }
 
 init {
    viewModelScope.launch(IO) {
        newsRepo
            .getArticles()
            .onEach { newArticle ->
                _articles.postValue(newArticle)
                Log.d("ARTICLE", newArticle.toString())
            }
            .collect()
    }
}
```

Then here you can do:

```suggestion
        fetchArticles()
```

And above, in the onCreate() method do this:

```kotlin
viewsModel.articles.observe(this) { result ->
    when(result) {
        is CustomResult.Success -> {
                articleAdapter.updateArticle(result.value)
        }
        is CustomResult.Failure -> {
            Toast.makeText(this@MainActivity, "No Data Available", Toast.LENGTH_LONG).show()
        }
    }
}
binding.swiperefresh.isRefreshing = false
```